### ISSUE TITLE
 Execution process

### ISSUE THREAD

**Issue by Zhongrocky** [creatis-myriad/plug-and-play-reco-regularization#2]
Nice work， I have a batch of 2D blood vessel prediction results, and I want to use this post-processing method to solve the discontinuity problem. Please tell me how to use this code. I understand the need to generate training data. Then I need to train the connection model based on the training data. Finally, I need to iteratively connect the models. However, I still don’t understand the order in which this code is used to complete my post-processing tasks.

**Comment by sophie-carneiro** [comment]
**Hello,**

Thank you for your question! Here's how you can use the code for your task:

1. **Functions for 2D Images:**  
   In the `sources/source_2D` directory, you'll find all the necessary functions for working with 2D images. The README describes what each file does.

2. **Pipeline Example:**  
   To understand the order of operations, check out the `example.py` file. It demonstrates how to:
   - Create a dataset.
   - Train the neural network.
   - Apply the trained model as a post-processing step.

3. **Pre-trained Model:**  
   If you don't have annotations or binary vascular trees to train a model, you can use the pre-trained model provided in the `modeles` directory.

Feel free to let me know if anything is unclear or if you encounter issues while setting it up!

**Comment by Zhongrocky** [comment]
I have a question, is the process of connecting 3D curves similar to the 2D process? I want to use this post-processing method to curve connect the results of my 3D vascular segmentation

**Comment by sophie-carneiro** [comment]
The overall pipeline (creating a dataset, training a model, and applying it iteratively) remains conceptually the same in 3D. However, the feature extraction, model design, and data representation will differ to accommodate the 3D structure. Therefore, you’ll find specific functions tailored for 3D data processing in the sources/source_3D directory and a similar `example.py` than in 2D.

**Comment by Zhongrocky** [comment]
@sophie-carneiro Hello, I encountered a problem during the debugging process as shown in the figure.
![image](https://github.com/user-attachments/assets/ae3d942d-d503-46d8-abcc-088f2c9ab907)

**Comment by sophie-carneiro** [comment]
It seems that in `train.py` (line 82), I set the random seed using` random.randint()`, which may occasionally exceed the maximum value of a number represented in uint32. I have updated the code in the repository by lowering the upper bound of the `random.randint()` function. However, since this issue does not occur on my setup, I cannot confirm with certainty that this change resolves the problem. Please let me know if the issue persists after this update.

**Comment by Zhongrocky** [comment]
I appreciate your help, but it didn't solve my problem

**Comment by Zhongrocky** [comment]
While running the code, our virtual environment may differ, resulting in different results.

**Comment by Zhongrocky** [comment]
@sophie-carneiro hello, I have a question. I want to know the reason why my loss curve doesn't decrease, as shown below：Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████| 43/43 [09:10<00:00, 12.79s/it]
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████| 11/11 [02:22<00:00, 13.00s/it]
----------
epoch 1/1000
1/11, train_loss: 1.0951
2/11, train_loss: 1.0760
3/11, train_loss: 0.9941
4/11, train_loss: 0.9952
5/11, train_loss: 0.9923
6/11, train_loss: 0.9992
7/11, train_loss: 0.9944
8/11, train_loss: 1.1473
9/11, train_loss: 0.9958
10/11, train_loss: 0.9997
11/11, train_loss: 0.9920
step :  11
saved new best metric model
current epoch: 1 current mean dice: 0.7266
best mean dice: 0.7266 at epoch: 1
----------
epoch 2/1000
1/11, train_loss: 0.9993
2/11, train_loss: 1.0622
3/11, train_loss: 0.9950
4/11, train_loss: 0.9964
5/11, train_loss: 0.9919
6/11, train_loss: 0.9999
7/11, train_loss: 1.0241
8/11, train_loss: 0.9970
9/11, train_loss: 0.9916
10/11, train_loss: 0.9924
11/11, train_loss: 0.9820
step :  11
saved new best metric model
current epoch: 2 current mean dice: 0.6852
best mean dice: 0.6852 at epoch: 2
----------
epoch 3/1000
1/11, train_loss: 0.9992
2/11, train_loss: 0.9990
3/11, train_loss: 0.9979
4/11, train_loss: 0.9935
5/11, train_loss: 0.9994
6/11, train_loss: 0.9890
7/11, train_loss: 0.9905
8/11, train_loss: 1.0833
9/11, train_loss: 1.1037
10/11, train_loss: 0.9988
11/11, train_loss: 0.9910
10/11, train_loss: 0.9988
11/11, train_loss: 0.9910
step :  11
current epoch: 3 current mean dice: 0.8107
best mean dice: 0.6852 at epoch: 2
----------
epoch 4/1000
1/11, train_loss: 0.9956
2/11, train_loss: 0.9910
3/11, train_loss: 0.9987
4/11, train_loss: 1.0912
5/11, train_loss: 0.9918
6/11, train_loss: 0.9992
7/11, train_loss: 1.1040
8/11, train_loss: 0.9966
9/11, train_loss: 0.9948
10/11, train_loss: 1.0484
11/11, train_loss: 0.9945
step :  11
saved new best metric model
current epoch: 4 current mean dice: 0.6641
best mean dice: 0.6641 at epoch: 4
----------
epoch 5/1000
1/11, train_loss: 0.9941
2/11, train_loss: 0.9949
3/11, train_loss: 0.9912
4/11, train_loss: 0.9952
5/11, train_loss: 1.1071
6/11, train_loss: 0.9997
7/11, train_loss: 0.9978
8/11, train_loss: 0.9987
9/11, train_loss: 1.0000
10/11, train_loss: 1.1069
11/11, train_loss: 1.0466
step :  11
current epoch: 5 current mean dice: 0.7745
best mean dice: 0.6641 at epoch: 4
----------
epoch 6/1000
1/11, train_loss: 0.9982
2/11, train_loss: 0.9938
3/11, train_loss: 0.9875
4/11, train_loss: 1.0936
5/11, train_loss: 1.0930
6/11, train_loss: 1.0972
7/11, train_loss: 0.9920
8/11, train_loss: 1.0000
9/11, train_loss: 0.9942
10/11, train_loss: 0.9917
11/11, train_loss: 0.9974
step :  11
current epoch: 6 current mean dice: 0.7004
best mean dice: 0.6641 at epoch: 4
----------
epoch 7/1000
1/11, train_loss: 0.9907
2/11, train_loss: 0.9925
3/11, train_loss: 1.0892
4/11, train_loss: 1.0755
5/11, train_loss: 0.9992
6/11, train_loss: 1.0620
7/11, train_loss: 0.9925
8/11, train_loss: 1.0893
9/11, train_loss: 1.1102
10/11, train_loss: 0.9979
11/11, train_loss: 0.9801
step :  11
saved new best metric model
current epoch: 7 current mean dice: 0.6637
best mean dice: 0.6637 at epoch: 7
----------
epoch 8/1000
1/11, train_loss: 0.9881
2/11, train_loss: 1.0723
3/11, train_loss: 0.9972
4/11, train_loss: 0.9989
5/11, train_loss: 0.9950
6/11, train_loss: 0.9956
7/11, train_loss: 0.9977
8/11, train_loss: 1.0000
9/11, train_loss: 0.9986
10/11, train_loss: 0.9970
11/11, train_loss: 1.0462
step :  11
current epoch: 8 current mean dice: 0.6813
best mean dice: 0.6637 at epoch: 7
----------
epoch 9/1000
1/11, train_loss: 0.9940
2/11, train_loss: 0.9904
3/11, train_loss: 1.1346
4/11, train_loss: 0.9988
5/11, train_loss: 1.0205
6/11, train_loss: 0.9992
7/11, train_loss: 0.9973
8/11, train_loss: 0.9951
9/11, train_loss: 0.9966
10/11, train_loss: 1.0000
11/11, train_loss: 1.0000
step :  11
current epoch: 9 current mean dice: 0.7900
best mean dice: 0.6637 at epoch: 7
----------
epoch 10/1000
1/11, train_loss: 1.0469
2/11, train_loss: 0.9995
3/11, train_loss: 0.9898
4/11, train_loss: 0.9940
5/11, train_loss: 0.9862
6/11, train_loss: 0.9966
7/11, train_loss: 0.9873
8/11, train_loss: 0.9840
9/11, train_loss: 0.9988
10/11, train_loss: 1.1207
11/11, train_loss: 0.9917
step :  11
current epoch: 10 current mean dice: 0.6977
best mean dice: 0.6637 at epoch: 7
----------
epoch 11/1000
1/11, train_loss: 1.2340
2/11, train_loss: 0.9980
3/11, train_loss: 0.9934
4/11, train_loss: 1.1167
5/11, train_loss: 0.9919
6/11, train_loss: 0.9991
7/11, train_loss: 1.0883
8/11, train_loss: 0.9915
9/11, train_loss: 0.9989
10/11, train_loss: 0.9947
11/11, train_loss: 1.2348
step :  11
saved new best metric model
current epoch: 11 current mean dice: 0.6621
best mean dice: 0.6621 at epoch: 11
----------
epoch 12/1000
1/11, train_loss: 1.0768
2/11, train_loss: 1.1060
3/11, train_loss: 1.0375
4/11, train_loss: 0.9949
5/11, train_loss: 1.0276
6/11, train_loss: 1.0000
7/11, train_loss: 0.9970
8/11, train_loss: 1.1062
9/11, train_loss: 0.9902
10/11, train_loss: 0.9986
11/11, train_loss: 1.3699
step :  11
current epoch: 12 current mean dice: 0.6762
best mean dice: 0.6621 at epoch: 11
----------
epoch 13/1000
1/11, train_loss: 0.9876
2/11, train_loss: 0.9916
3/11, train_loss: 0.9923
4/11, train_loss: 0.9986
5/11, train_loss: 1.2490
6/11, train_loss: 1.0000
7/11, train_loss: 0.9902
8/11, train_loss: 0.9831
9/11, train_loss: 1.1286
10/11, train_loss: 1.0000
11/11, train_loss: 1.1803
step :  11
current epoch: 13 current mean dice: 0.6843
best mean dice: 0.6621 at epoch: 11
----------
epoch 14/1000
1/11, train_loss: 0.9934
2/11, train_loss: 0.9940
3/11, train_loss: 0.9825
4/11, train_loss: 1.1300
5/11, train_loss: 0.9978
6/11, train_loss: 1.0787
7/11, train_loss: 1.0856
8/11, train_loss: 1.1014
9/11, train_loss: 0.9968
10/11, train_loss: 0.9955
11/11, train_loss: 0.9938
step :  11
current epoch: 14 current mean dice: 0.6808
best mean dice: 0.6621 at epoch: 11
----------
epoch 15/1000
1/11, train_loss: 0.9945
2/11, train_loss: 1.0821
3/11, train_loss: 0.9887
4/11, train_loss: 0.9901
5/11, train_loss: 0.9980
6/11, train_loss: 1.1102
7/11, train_loss: 0.9841
8/11, train_loss: 0.9929
9/11, train_loss: 1.0634
10/11, train_loss: 1.0270
11/11, train_loss: 1.1170
step :  11
current epoch: 15 current mean dice: 0.6632
best mean dice: 0.6621 at epoch: 11
----------
epoch 16/1000
1/11, train_loss: 0.9921
2/11, train_loss: 0.9959
3/11, train_loss: 1.1323
4/11, train_loss: 1.0900
5/11, train_loss: 0.9980
6/11, train_loss: 0.9894
7/11, train_loss: 0.9968
8/11, train_loss: 1.0716
9/11, train_loss: 1.1637
10/11, train_loss: 1.0873
11/11, train_loss: 1.0000
step :  11
current epoch: 16 current mean dice: 0.6866
best mean dice: 0.6621 at epoch: 11
----------
epoch 17/1000
1/11, train_loss: 1.0000
2/11, train_loss: 0.9997
3/11, train_loss: 0.9933
4/11, train_loss: 0.9935
5/11, train_loss: 0.9956
6/11, train_loss: 0.9938
7/11, train_loss: 0.9927
8/11, train_loss: 1.0738
9/11, train_loss: 0.9980
10/11, train_loss: 0.9986
11/11, train_loss: 0.9906
step :  11
current epoch: 17 current mean dice: 0.6682
best mean dice: 0.6621 at epoch: 11
----------
epoch 18/1000
1/11, train_loss: 1.3917
2/11, train_loss: 1.0980
3/11, train_loss: 1.0227
4/11, train_loss: 0.9874
5/11, train_loss: 1.0000
6/11, train_loss: 0.9899
7/11, train_loss: 1.0055
8/11, train_loss: 0.9994
9/11, train_loss: 0.9933
10/11, train_loss: 0.9999
11/11, train_loss: 1.0000
step :  11
current epoch: 18 current mean dice: 0.7109
best mean dice: 0.6621 at epoch: 11
----------
epoch 19/1000
1/11, train_loss: 0.9926
2/11, train_loss: 0.9994
3/11, train_loss: 1.0409
4/11, train_loss: 1.0873
5/11, train_loss: 0.9989
6/11, train_loss: 0.9982
7/11, train_loss: 0.9905
8/11, train_loss: 0.9995
9/11, train_loss: 0.9942
10/11, train_loss: 1.0000
11/11, train_loss: 1.0850
step :  11
current epoch: 19 current mean dice: 0.6696
best mean dice: 0.6621 at epoch: 11

**Comment by sophie-carneiro** [comment]
Hello, 
there were a little mistake with the computing of the validation loss. I corrected it. I used to divide by the number of element at each batch and not at each epoch . However, it should not have an impact on the training. 
With 19 epochs, I think it is hard to see the evolution of the training. You should try longer training. If it is still not converging, you can play with the hyperparameters of the training of the function `train()`, or even add data for training.

**Comment by Zhongrocky** [comment]
@sophie-carneiro thank you
